d2/src/main/java/com/linkedin/d2/xds/XdsClient.java

```diff
@@ -0,0 +1,280 @@
+package com.linkedin.d2.xds;
+
+import com.google.protobuf.Value;
+import io.envoyproxy.envoy.config.endpoint.v3.LocalityLbEndpoints;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+public abstract class XdsClient
+{
+  /**
+   * Watcher interface for a single requested xDS resource.
+   */
+  interface ResourceWatcher
+  {
+    // TODO: Handle onError and onResourceNotExist
+  }
+
+  interface LdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(LdsUpdate update);
+  }
+
+  interface RdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(RdsUpdate update);
+  }
+
+  interface CdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(CdsUpdate update);
+  }
+
+  interface EdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(EdsUpdate update);
+  }
+
+  interface ResourceUpdate
+  {
+  }
+
+  static final class LdsUpdate implements ResourceUpdate
+  {
+    // Http level api listener configuration.
+    final HttpConnectionManager _httpConnectionManager;
+
+    LdsUpdate(HttpConnectionManager httpConnectionManager)
+    {
+      _httpConnectionManager = httpConnectionManager;
+    }
+
+    public HttpConnectionManager getHttpConnectionManager()
+    {
+      return _httpConnectionManager;
+    }
+
+    @Override
+    public String toString()
+    {
+      return "LdsUpdate{" + "_httpConnectionManager=" + _httpConnectionManager + '}';
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(_httpConnectionManager);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o)
+      {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass())
+      {
+        return false;
+      }
+      LdsUpdate ldsUpdate = (LdsUpdate) o;
+      return Objects.equals(_httpConnectionManager, ldsUpdate._httpConnectionManager);
```

Comment:
Not sure about this: in Golang at least, two Protobufs which have the same public fields can fail the DeepEquals() check because they may have different private fields, and to make things worse, the private fields are generated by protobuf deserialization libraries. Could you verify the Java behavior (even with debuggers)?

Reply:
https://developers.google.com/protocol-buffers/docs/reference/go/faq#deepequal

Reply:
From online docs, seems like Java .equals() API behaves well